Use switch statement for bufferOwnership inside StringImpl destructor

cdumez opened the pull request:
#### d2682a1e8a99dfbb23fb380e86b5fabba6132dd2
<pre>
Use switch statement for bufferOwnership inside StringImpl destructor
<a href="https://bugs.webkit.org/show_bug.cgi?id=253735">https://bugs.webkit.org/show_bug.cgi?id=253735</a>

Reviewed by Darin Adler.

Use switch statement for bufferOwnership inside StringImpl destructor, instead
of multiple if checks.

* Source/WTF/wtf/text/StringImpl.cpp:
(WTF::StringImpl::~StringImpl):

Canonical link: <a href="https://commits.webkit.org/261542@main">https://commits.webkit.org/261542@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c1a06d13a886114447c7f511d541f171f0feeb12

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/111976 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/21104 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/589 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/3740 "Built successfully") | [❌ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/120638 "Failed to compile WebKit") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/22453 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/12252 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/3486 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/117736 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/16672 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/99834 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/104997 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/98624 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/394 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/45655 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/100416 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/13534 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/412 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/93054 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/11695 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/14209 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/9818 "5 flakes 3 failures") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/101694 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/19583 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52416 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/31587 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8045 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/16016 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/109736 "Built successfully") | 
| | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/26992 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WTF/wtf/text/StringImpl.cpp:
```diff
@@ -129,26 +129,25 @@ StringImpl::~StringImpl()
         external->freeExternalBuffer(const_cast<LChar*>(m_data8), sizeInBytes());
         external->m_free.~ExternalStringImplFreeFunction();
-        return;
+        break;
```
can all these breaks be returns instead?

cdumez replied:
They could. Why would that be better?

rr-codes replied:
To me personally it's more readable since it implies there's no more logic after the switch-case but ya it's not objectively better 😄

cdumez committed:
Use switch statement for bufferOwnership inside StringImpl destructor
https://bugs.webkit.org/show_bug.cgi?id=253735

Reviewed by Darin Adler.

Use switch statement for bufferOwnership inside StringImpl destructor, instead
of multiple if checks.

* Source/WTF/wtf/text/StringImpl.cpp:
(WTF::StringImpl::~StringImpl):

Canonical link: https://commits.webkit.org/261542@main
